### ISSUE TITLE
The unobserved exception is fired once operation timeout is reached

### ISSUE THREAD

**Issue by AdamMachera** [Azure/azure-iot-sdk-csharp#265]
Windows 10
DeviceClient 1.5.2

If I send data to IotHub and at some point enable firewall rule that blocks AMQP ports I'm getting following exception

Message val-0 was sent
Message val-1 was sent
Message val-2 was sent
Message val-3 was sent
Message val-4 was sent
Message val-5 was sent
Message val-6 was sent
Message val-7 was sent
System.TimeoutException: Operation timeout expired.
   at Microsoft.Azure.Devices.Client.DeviceClient.<>c__DisplayClass64_0.<ApplyTimeout>b__2(Task t)
   at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at ConsoleApp3.Program.<>c__DisplayClass0_0.<<Main>b__1>d.MoveNext() in C:\Users\amachera\source\repos\ConsoleApp3\ConsoleApp3\Program.cs:line 49
System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'FaultTolerantAmqpObject`1'.
   at Microsoft.Azure.Devices.Client.Singleton`1.<GetOrCreateAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.IotHubConnection.<GetSessionAsync>d__21.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.IotHubConnection.<CreateSendingLinkAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Transport.AmqpTransportHandler.<CreateEventSendingLinkAsync>d__63.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.FaultTolerantAmqpObject`1.<OnCreateAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Singleton`1.<CreateValue>d__19.MoveNext()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'FaultTolerantAmqpObject`1'.
   at Microsoft.Azure.Devices.Client.Singleton`1.<GetOrCreateAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.IotHubConnection.<GetSessionAsync>d__21.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.IotHubConnection.<CreateSendingLinkAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Transport.AmqpTransportHandler.<CreateEventSendingLinkAsync>d__63.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.FaultTolerantAmqpObject`1.<OnCreateAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Singleton`1.<CreateValue>d__19.MoveNext()<---

I created simple application that generates data and using TPL DataBlock I'm consuming data using 3 tasks.
Unobserved exception is triggered once DeviceClient operation timeout is reached.
SendEventAsync operation is synchronized with SemphoreSlim.

[ConsoleApp3.zip](https://github.com/Azure/azure-iot-sdk-csharp/files/1526600/ConsoleApp3.zip)

**Comment by CIPop** [comment]
Thanks for reporting @AdamMachera. I've tagged this to point out that further investigation is needed to root-cause the issue and that we welcome external contributor support.

**Comment by FilipVanRaemdonck** [comment]
I do have the same behaviour, and it's very difficult to catch this error. I there a fix for this error already?

**Comment by CIPop** [comment]
@FilipVanRaemdonck I've tried reproing this with 1.6.1 and [Clumsy](http://jagt.github.io/clumsy/) (set to Drop 100% for port 5671) and got the exception on the right Task completion and the app was able to continue several times:

```
Message val-274 was sent
Message val-275 was sent
Message val-276 was sent
Message val-277 was sent
Message val-278 was sent
Message val-279 was sent
Message val-280 was sent
Message val-281 was sent
System.TimeoutException: Operation timeout expired.
   at Microsoft.Azure.Devices.Client.DeviceClient.<>c.<ApplyTimeout>b__73_2(Task t)
   at System.Threading.Tasks.ContinuationTaskFromTask.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at ConsoleApp3.Program.<>c__DisplayClass0_0.<<Main>b__1>d.MoveNext() in C:\Users\crispop\Desktop\cust\ConsoleApp3\Program.cs:line 49
Message val-283 was sent
Message val-284 was sent
Message val-285 was sent
Message val-286 was sent
Message val-287 was sent
```

I've modified the code above to ensure I'm not getting an UnobservedTaskException:

```C#
        private static void TaskScheduler_UnobservedTaskException(object sender, UnobservedTaskExceptionEventArgs e)
        {
            Console.WriteLine("BUG!" + e.Exception);
        }
```

@AdamMachera @FilipVanRaemdonck Can you still repro using 1.6.1?

**Comment by AdamMachera** [comment]
it is still happening in version 1.6.1 if I disable traffic on AMQP ports.
I was using this rule to block traffic to IotHub and didn't modify connection timeout to IotHub

![image](https://user-images.githubusercontent.com/9037341/34826099-87951d06-f6d5-11e7-925d-0365a170a3b7.png)

**Comment by FilipVanRaemdonck** [comment]
We still have the same problem

**Comment by jasmineymlo** [comment]
@AdamMachera @FilipVanRaemdonck, Just an update that I was able to a repro on my machine by disabling traffic on AMQP ports with debug build of the latest master build.  Not seeing a repro with debugger.  Will continue with investigation.  Thanks.

**Comment by jasmineymlo** [comment]
I believe that the unobserved exceptions are originated from 
await Task.WhenAny(a,b);
The above will ultimately throw an exceptions from whichever task faults first.  In our case, when operation timeout.  The exception will bubble up and caught by the app.
Then the other task, which is trying to connect in the network layer, fails and throw socket exception which is unobserved.  Currently, we do not have a way to “cancel” the outstanding tasks which is trying to open the connection.

AFAIK, since .NET4.5, unobserved exceptions still raise the UnobservedTaskException event, but the process does not terminate.
https://msdn.microsoft.com/en-us/library/system.threading.tasks.taskscheduler.unobservedtaskexception(v=vs.110).aspx

Base on the above, we could safely ignore the unobserved exceptions which are throw from the socket and the main app should not be affected.  

I understand that this is not ideal.  We do have plans to refactor our operation timeout code path to better handle the cancellation of tasks.  These refactoring/redesigning will change the behavior on timeout operations.  Thanks.

**Comment by CIPop** [comment]
> We do have plans to refactor our operation timeout code path to better handle the cancellation of tasks. These refactoring/redesigning will change the behavior on timeout operations. Thanks.

Since the exceptions don't crash the process and are there because we are not observing/awaiting tasks that either timed out or otherwise failed, we will postpone the fix for a later time.

**Comment by victor-mandujano** [comment]
I'm still able to reproduce using Microsoft.Azure.Devices.Client(1.37.2)
1. When blocking TCP traffic on MQTT port 8883 I get repeated unobserved task exceptions with an inner exception of type DotNetty.Transport.Channels.ConnectTimeoutException
2. When there are transient network failures I get the same with an inner exception of type System.TimeoutException

Simply catching and ignoring **these kind** of unobserved task exceptions seems to be an ugly hack in certain scenarios when people needs to actually deal with other kinds of unobserved tasks exceptions arising from their own internal libraries, etc.

Any plans to fix this?
System.TimeoutException: Connection hasn't been established in time.\n   at Microsoft.Azure.Devices.Client.Transport.Mqtt.MqttTransportHandler.OpenInternalAsync(CancellationToken cancellationToken)
